### PR TITLE
cosmos spark3 adds supports for shading 

### PR DESCRIPTION
**Pull request by moderakh**
This PR adds support for shading jar, I tested the jar in databricks and it works.

- builds an uber jar with all dependencies
- relocate the dependencies to ensure there won't be any dependency conflict with any other packages in the classpath

bugfix for ImplementationBridgeHelpers
 fixes a bug in ImplementationBridgeHelpers to ensure the referenced classes are loaded before accessors are referenced.

TODO: the jar with dependencies distribution legal aspect: https://github.com/Azure/azure-sdk-for-java/issues/19028
TODO: I need to verify that the optimization netty libs (native netty jar boringssl and netty native epoll) (which are already packaged in the jar) are utilized by netty

### CODE REVIEW

**Comment by mitchdenny** [comment]
@alzimmermsft thoughts on shading?

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml [hunk 1]
```diff
@@ -279,6 +280,151 @@
...
+            </goals>
+            <configuration>
+              <relocations>
+                <relocation>
+                  <pattern>com.azure</pattern>
+                  <shadedPattern>${shadingPrefix}.com.azure</shadedPattern>
+                  <excludes>
+                    <exclude>com.azure.cosmos.spark.*</exclude>
+                    <exclude>com.azure.cosmos.implementation.SparkBridgeInternal</exclude>
+                    <exclude>com.azure.cosmos.models.CosmosParametrizedQuery</exclude>
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml:302]: NIT: typo in the class name (unrelated, definitely not blocking)

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml [hunk 2]
```diff
@@ -279,6 +280,151 @@
           <skip>true</skip>
         </configuration>
       </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
+        <version>3.2.4</version> <!-- {x-version-update;org.apache.maven.plugins:maven-shade-plugin;external_dependency} -->
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml:287]: Nice - this dependency is used already and has been added to external dependencies :-)

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml [hunk 3]
```diff
@@ -25,6 +25,7 @@
     <maven.compiler.source>11</maven.compiler.source>
     <maven.compiler.target>11</maven.compiler.target>
     <jacoco.skip.coverage.check>true</jacoco.skip.coverage.check>
+    <shadingPrefix>azure_cosmos_spark</shadingPrefix>
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml:28]: Should we have some versioning in the shading prefix (would it help to support multiple versions/deployments) - don't know just a question. Not blocking.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-spark_3-0_2-12/pom.xml:302]: Thanks typo fixed.

**Pull request merged**